### ISSUE TITLE
[libc] Clean up unnecessary function pointers in scanf

### ISSUE THREAD

**Issue by michaelrj-google** [llvm/llvm-project#115394]
The scanf reader is the abstraction that scanf uses to read from either a file or a string, similar to printf's writer. Unlike printf, however, there are only two variants of scanf: `fscanf` and `sscanf` (just `scanf` is the same as `fscanf(stdin, ...)`). This means that the `stream_getc` and `stream_ungetc` functions don't need to be pointers, they can be declared with `using` statements based on the current platform.

Files you'll need to edit:
[libc/src/stdio/scanf_core/vfscanf_internal.h](https://github.com/llvm/llvm-project/blob/main/libc/src/stdio/scanf_core/vfscanf_internal.h)
[libc/src/stdio/scanf_core/reader.h](https://github.com/llvm/llvm-project/blob/main/libc/src/stdio/scanf_core/reader.h)

What to do:
Move the definitions of `getc` and `ungetc` in `vfscanf_internal` into `reader.h` (along with the associated `#if defined` logic for each target).
Define the `stream_getc` and `stream_ungetc` functions with `using` statements inside of `class Reader` (or do something equivalent).
Clean up the relevant code.

If you have questions or need help, feel free to comment on this issue or reach out via the LLVM discord.

**Comment by AidanGoldfarb** [comment]
@shawbyoung are you still working on this issue? If not I would be happy to take over

**Comment by shawbyoung** [comment]
@AidanGoldfarb Yes, I was busy but now am carving out some time to put out a PR in the next few days.

**Comment by vinay-deshmukh** [comment]
@shawbyoung are you still working on this issue?

**Comment by vinay-deshmukh** [comment]
@michaelrj-google 

Is it currently possible to build `llvm-libc` in overlay mode on macos/m1?

My code (in the linked PR) is failing on 2 issues right now:

1. `Mutex` is not recognized as a type, https://github.com/llvm/llvm-project/blob/3496e96f78c46f5b94c1892f97c470fd89293795/libc/src/__support/File/file.h#L90
2. Macro substitutions that happen for: `clearerr_unlocked`, which is a macro in macos' `stdio.h`, but defined as a method name in https://github.com/llvm/llvm-project/blob/3496e96f78c46f5b94c1892f97c470fd89293795/libc/src/__support/File/file.h#L260

both errors happen in the CI as well: https://github.com/llvm/llvm-project/pull/121215#issuecomment-2564379404

Would appreciate any guidance on solving the above issues as well.

**Comment by michaelrj-google** [comment]
The MacOS build is working for me right now. Here's my build process:
1) Delete CMakeCache.txt (MacOS likes to move around paths to headers after updates)
2) Run my cmake command: `cmake ../runtimes -G Ninja -DLLVM_ENABLE_RUNTIMES="libc"`
3) Run the tests: `ninja check-libc`

This works on my M1 Mac as of f9369cc602272796c15de1065a782f812e791df3
